### PR TITLE
Add support for CSE Log mapping

### PR DESCRIPTION
**Pull request by pmontiel-sumo**

### CODE REVIEW

File: sumologic/resource_sumologic_cse_log_mapping.go [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+	d.Set("relates_entities", CSELogMapping.RelatesEntities)
+	d.Set("skipped_values", CSELogMapping.SkippedValues)
+	setFields(d, CSELogMapping.Fields)
+	setStructuredInputs(d, CSELogMapping.StructuredInputs)
+	if CSELogMapping.UnstructuredFields != nil {
+		setUnStructuredFields(d, *(CSELogMapping.UnstructuredFields))
+	}
+}
+
+func setUnStructuredFields(d *schema.ResourceData, unstructuredFields CSELogMappingUnstructuredFields) {
```
Comment (josh-williams) [inline sumologic/resource_sumologic_cse_log_mapping.go:423]: Minor nit: The "S" in "Unstructured" should be lowercase.

File: website/docs/r/cse_log_mapping.html.markdown [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+```
+
+## Argument reference
+
+The following arguments are supported:
+
+- `name` - (Required) The name of the log mapping.
+- `parent_id` - (Optional) The id of the parent log mapping.
+- `product_guid` - (Required) Product GUID.
+- `record_type` - (Required) The record type to be created. ( possible values: Audit, AuditChange, AuditFile, AuditResourceAccess, Authentication, AuthenticationPrivilegeEscalation, Canary, Email, Endpoint, EndpointModuleLoad, EndpointProcess, Network, NetworkDHCP, NetworkDNS, NetworkFlow, NetworkHTTP, NetworkProxy, Notification, NotificationVulnerability )
```
Comment (josh-williams) [inline website/docs/r/cse_log_mapping.html.markdown:56]: Can we remove the extra space after the open parentheses and before the close parentheses here?

File: website/docs/r/cse_log_mapping.html.markdown [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+- `enabled` - (Required) Enabled flag.
+- `relates_entities` - (Optional) Set to true to relate entities.
+- `skipped_values` - (Optional) List of skipped values.
+- `fields` - (Required) List of fields for the new log mapping. See [field_schema](#schema-for-field) for details.
+- `structured_inputs` - (Optional, omit if unstructured_fields is defined ) List of structured inputs for the new log mapping. See [structured_input_schema](#schema-for-structured_input) for details.
```
Comment (josh-williams) [inline website/docs/r/cse_log_mapping.html.markdown:61]: Let's remove the extra space before the close parentheses here.

File: website/docs/r/cse_log_mapping.html.markdown [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+- `product` - (Required) Product name.
+- `vendor` - (Required) Vendor name.
+
+### Schema for `unstructured_field`
+- `pattern_names` - (Required) List of grok pattern names.
+
+
+The following attributes are exported:
+
+- `id` - The internal ID of the rule tuning expression.
```
Comment (josh-williams) [inline website/docs/r/cse_log_mapping.html.markdown:96]: The ID of the log mapping

**Review by josh-williams — COMMENTED** [review]
LGTM. Just a few comments inline.

File: sumologic/resource_sumologic_cse_log_mapping.go (hunk 1 above)
Comment (pmontiel-sumo) [reply sumologic/resource_sumologic_cse_log_mapping.go]: done under:
https://github.com/SumoLogic/terraform-provider-sumologic/pull/284/commits/a994b2d8d74888aad0fa04698f1dd7d8084a3ce6

File: website/docs/r/cse_log_mapping.html.markdown (hunk 3 above)
Comment (pmontiel-sumo) [reply website/docs/r/cse_log_mapping.html.markdown]: done under 
https://github.com/SumoLogic/terraform-provider-sumologic/pull/284/commits/a994b2d8d74888aad0fa04698f1dd7d8084a3ce6

File: website/docs/r/cse_log_mapping.html.markdown (hunk 4 above)
Comment (pmontiel-sumo) [reply website/docs/r/cse_log_mapping.html.markdown]: done under:
https://github.com/SumoLogic/terraform-provider-sumologic/pull/284/commits/a994b2d8d74888aad0fa04698f1dd7d8084a3ce6

File: sumologic/resource_sumologic_cse_log_mapping.go [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+				},
+			},
+			"structured_inputs": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: getLogMappingStructuredInputFieldSchema(),
+				},
+			},
+			"unstructured_fields": {
```
Comment (sumovishal) [inline sumologic/resource_sumologic_cse_log_mapping.go:151]: `structuredFields` is missing here. Is that field not supported?

File: sumologic/data_source_sumologic_cse_log_mapping_vendor_product.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+	vendorAndProduct, err := c.GetCSELogMappingVendorsAndProducts(d.Get("product").(string), d.Get("vendor").(string))
+
+	if err != nil {
+		return err
+	}
+
+	if vendorAndProduct == nil {
+		d.SetId("")
+		return fmt.Errorf("Vendor product not found, removing from state: %v - %v", id, err)
```
Comment (sumovishal) [inline sumologic/data_source_sumologic_cse_log_mapping_vendor_product.go:44]: `err` will be nil here. Remove it from the error message?

File: sumologic/sumologic_cse_log_mapping_vendor_product.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+)
+
+func (s *Client) GetCSELogMappingVendorsAndProducts(product, vendor string) (*CSELogMappingVendorAndProduct, error) {
+
+	data, _, err := s.Get(fmt.Sprintf("sec/v1/vendors-and-products?product=%s&vendor=%s", url.QueryEscape(product), url.QueryEscape(vendor)))
```
Comment (sumovishal) [inline sumologic/sumologic_cse_log_mapping_vendor_product.go:11]: I don't see this endpoint in API docs https://api.sumologic.com/docs/sec. Are we planning to release this API?

File: sumologic/resource_sumologic_cse_log_mapping_test.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+)
+
+func TestAccSumologicSCELogMapping_create(t *testing.T) {
```
Comment (sumovishal) [inline sumologic/resource_sumologic_cse_log_mapping_test.go:11]: Add an acceptance test to update log mapping?

File: sumologic/data_source_sumologic_cse_log_mapping_vendor_product.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
+package sumologic
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func dataSourceCSELogMappingVendorAndProduct() *schema.Resource {
```
Comment (sumovishal) [inline sumologic/data_source_sumologic_cse_log_mapping_vendor_product.go:8]: We need an acceptance test for this data source.

**Review by sumovishal — COMMENTED** [review]
Left some comments.

Can you remove `website/docs/README.md`? We no longer use it.

File: sumologic/resource_sumologic_cse_log_mapping.go (hunk 5 above)
Comment (pmontiel-sumo) [reply sumologic/resource_sumologic_cse_log_mapping.go:151]: it is a deprecated field, structured_inputs usage instead

File: sumologic/data_source_sumologic_cse_log_mapping_vendor_product.go (hunk 6 above)
Comment (pmontiel-sumo) [reply sumologic/data_source_sumologic_cse_log_mapping_vendor_product.go]: done under https://github.com/SumoLogic/terraform-provider-sumologic/pull/284/commits/a44f995f197a0d27ca41d773aa10652ebeba4e48

File: sumologic/sumologic_cse_log_mapping_vendor_product.go (hunk 7 above)
Comment (pmontiel-sumo) [reply sumologic/sumologic_cse_log_mapping_vendor_product.go:11]: currently is defined as private, but is already there ( being used in acceptance test  )

File: sumologic/data_source_sumologic_cse_log_mapping_vendor_product.go (hunk 9 above)
Comment (pmontiel-sumo) [reply sumologic/data_source_sumologic_cse_log_mapping_vendor_product.go:8]: https://github.com/SumoLogic/terraform-provider-sumologic/pull/284/commits/a44f995f197a0d27ca41d773aa10652ebeba4e48

File: website/docs/r/cse_log_mapping.html.markdown [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+- `structured_inputs` - (Optional, omit if unstructured_fields is defined) List of structured inputs for the new log mapping. See [structured_input_schema](#schema-for-structured_input) for details.
+- `unstructured_fields` - (Optional, omit if structured_inputs is defined) Unstructured fields for the new log mapping. See [unstructured_field_schema](#schema-for-unstructured_field) for details.
+
+### Schema for `field`
+- `name` - (Required) Name of the field.
+- `value` - (Optional) Value of the field.
+- `value_type` - (Optional) The value type.
+- `skipped_values` - (Optional) List of skipped values.
+- `default_value` - (Optional) Default value of the field.
+- `format` - (Optional) Format of the field. ( JSON, Windows, Syslog, CEF, LEEF )
```
Comment (josh-williams) [inline website/docs/r/cse_log_mapping.html.markdown:70]: Nit: extra space after open parens and before close parens.

**Review by josh-williams — COMMENTED** [review]
LGTM. Just a few minor comments on the docs.

File: website/docs/r/cse_log_mapping.html.markdown [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+- `format_parameters` - (Optional) List of format parameters.
+- `lookup` - (Optional) List of lookup key value pair for field. See [lookup_schema](#schema-for-lookup) for details.
+
+### Schema for `lookup`
+- `key` - (Required) Lookup key.
+- `value` - (Required) Lookup value.
+
+### Schema for `structured_input`
+- `event_id_pattern` - (Required) Event id pattern.
+- `log_format` - (Required) Log format. ( JSON, Windows, Syslog, CEF, LEEF )
```
[the remark this answers is not in the document]
Comment (pmontiel-sumo) [reply website/docs/r/cse_log_mapping.html.markdown]: done under https://github.com/SumoLogic/terraform-provider-sumologic/pull/284/commits/a429a460c76c8f725c8c2d4e1441a5f9bff1a8f4

File: website/docs/r/cse_log_mapping.html.markdown (hunk 10 above)
Comment (pmontiel-sumo) [reply website/docs/r/cse_log_mapping.html.markdown]: done under https://github.com/SumoLogic/terraform-provider-sumologic/pull/284/commits/a429a460c76c8f725c8c2d4e1441a5f9bff1a8f4

File: website/docs/r/cse_log_mapping.html.markdown [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+- `name` - (Required) Name of the field.
+- `value` - (Optional) Value of the field.
+- `value_type` - (Optional) The value type.
+- `skipped_values` - (Optional) List of skipped values.
+- `default_value` - (Optional) Default value of the field.
+- `format` - (Optional) Format of the field. ( JSON, Windows, Syslog, CEF, LEEF )
+- `case_insensitive` - (Optional) Case insensitive flag.
+- `alternate_values` - (Optional) List of alternate values.
+- `time_zone` - (Optional) Time zone.
+- `split_delimiter` - (Optional) Split delimiter to be used. ( some example: ",", "-", "|")
```
[the remark this answers is not in the document]
Comment (pmontiel-sumo) [reply website/docs/r/cse_log_mapping.html.markdown]: done under https://github.com/SumoLogic/terraform-provider-sumologic/pull/284/commits/a429a460c76c8f725c8c2d4e1441a5f9bff1a8f4

File: website/docs/r/cse_log_mapping.html.markdown [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+  relates_entities = "true"
+  skipped_values = ["skipped"]
+  fields {
+    name = "action"
+    value = "action"
+    value_type = "constant"
+    skipped_values = ["-"]
+    default_value = ""
+    format = "JSON"
+    case_insensitive = "false"
```
[the remark this answers is not in the document]
Comment (pmontiel-sumo) [reply website/docs/r/cse_log_mapping.html.markdown]: done under https://github.com/SumoLogic/terraform-provider-sumologic/pull/284/commits/a429a460c76c8f725c8c2d4e1441a5f9bff1a8f4

**Review by sumovishal — APPROVED** [review]

**Pull request merged**